src/include/storage/record_buffer.h

```diff
@@ -32,22 +32,150 @@ class BufferSegment {
+  BufferSegment *Reset() {
+    size_ = 0;
+    return this;
+  }
 
  private:
+  template <class RecordType>
+  friend class IterableBufferSegment;
+
   friend class UndoBuffer;
+
   byte bytes_[common::Constants::BUFFER_SEGMENT_SIZE];
-  uint32_t end_ = 0;
+  uint32_t size_ = 0;
 };
 
+/**
+ * A thin wrapper around a buffer segment to allow iteration through its contents as the given template type
+ * @tparam RecordType records are treated as this type when iterating. Must expose an instance method Size() that
+ *                    gives the size of the record in memory in bytes
+ */
+template <class RecordType>
+class IterableBufferSegment {
+ public:
+  /**
+   * Iterator for iterating through the records in the buffer
+   */
+  class Iterator {
+   public:
+    /**
+     * @return reference to the underlying record
+     */
+    RecordType &operator*() const { return *reinterpret_cast<RecordType *>(segment_->bytes_ + segment_offset_); }
+
+    /**
+     * @return pointer to the underlying record
+     */
+    RecordType *operator->() const { return reinterpret_cast<RecordType *>(segment_->bytes_ + segment_offset_); }
+
+    /**
+     * prefix increment
+     * @return self-reference
+     */
+    Iterator &operator++() {
+      RecordType &me = this->operator*();
+      segment_offset_ += me.Size();
+      return *this;
+    }
+
+    /**
+     * postfix increment
+     * @return iterator that is equal to this before increment
+     */
+    const Iterator operator++(int) {
+      Iterator copy = *this;
+      operator++();
+      return copy;
+    }
+
+    /**
+     * equality check
+     * @param other the other iterator to compare to
+     * @return if the two iterators point to the same underlying record
+     */
+    bool operator==(const Iterator &other) const {
+      return segment_offset_ == other.segment_offset_ && segment_ == other.segment_;
+    }
+
+    /**
+     * inequality check
+     * @param other the other iterator to comapre to
+     * @return if the two iterators point to different underlying records
+     */
+    bool operator!=(const Iterator &other) const { return !(*this == other); }
+
+   private:
+    friend class IterableBufferSegment;
+    Iterator(BufferSegment *segment, uint32_t segment_offset) : segment_(segment), segment_offset_(segment_offset) {}
+    BufferSegment *segment_;
+    uint32_t segment_offset_;
+  };
+
+  /**
+   * Instantiates an IterableBufferSegment as a wrapper around a buffer segment
+   * @param segment
+   */
+  explicit IterableBufferSegment(BufferSegment *segment) : segment_(segment) {}
+
+  /**
+   * @return iterator to the first element
+   */
+  Iterator begin() { return {segment_, 0}; }
+
+  /**
+   * @return iterator to the second element
+   */
+  Iterator end() { return {segment_, segment_->size_}; }
+
+ private:
+  BufferSegment *segment_;
+};
+
+/**
+ * Custom allocator used for the object pool of buffer segments
+ */
+class RecordBufferSegmentAllocator {
```

Comment:
Maybe this class should go to common/allocator.h?

Reply:
I think a specialized one like this lives with the typedef for the object pool